pkg/storageos/deploy.go

```diff
@@ -1087,6 +1089,14 @@ func (s *Deployment) addNodeAffinity(podSpec *v1.PodSpec) {
 	}
 }
 
+// addToleration adds toleration to the given pod spec from the cluter
+// spec Toleration.
+func (s *Deployment) addToleration(podSpec *v1.PodSpec) {
+	if len(s.stos.Spec.Tolerations) > 0 {
+		podSpec.Tolerations = s.stos.Spec.Tolerations
+	}
```

Comment:
I think we should add a check here to ensure that `Toleration.Value` is not set if `Toleration.Operator` is `TolerationOpExists`. Passing a value with exists makes the configuration invalid. This would make this function to return an error. An error check should be added to the `addToleration()` call in `createDaemonSet()`. With that, you be able to test this by adding a test case in the `TestDeployTolerations` in deploy_test.go where you set value with operator exists. Deployment.Deploy() would return an error.

Before adding this, can you check if adding the invalid configuration fails at `createOrUpdateObject()` call at the end of `createDaemonSet()`. If it does, and a proper error is visible in the description of the storageoscluster object events, maybe we need not add this check.

Reply:
Looks like this is a bit of documentation lack from kubernetes site. As in the documentation is stated:
> the operator is Exists (in which case no value should be specified)

and the sourcecode show that is should work.

https://github.com/kubernetes/kubernetes/blob/a9fd9cef765a0b6d3854b187dcb6d05097f50c6a/staging/src/k8s.io/api/core/v1/toleration.go#L37-L56

But I found that the validation checks this:

https://github.com/kubernetes/kubernetes/blob/f7c4389b793cd6cf0de8d67f2c5db28b3985ad59/pkg/apis/core/validation/validation.go#L2813-L2855

I've tried to run the tests with `TolerationOpExists` and it does not fail, therefore I will implement the check early on.